[FLINK-25851][Connectors][Cassandra][test] support dynamic tables names for POJOs

echauchot opened the pull request:
R: @zentol

zentol reviewed (commented):
While I get the idea behind the change I'm not a fan of it.
It inherently means that we start out in a dirty state, which may or may not cause issues down the line.

zentol commented:
What we could instead look into is changing the annotation parameters of the POJOs (keyspace, table) via Reflection so each test uses a unique table.

echauchot commented:
> While I get the idea behind the change I'm not a fan of it. It inherently means that we start out in a dirty state, which may or may not cause issues down the line.

I agree. I did not like this approach either but I must admit I did not have any other idea because I could not see why we ended up having an already existing table despite all the cleaning measures.

echauchot commented:
> What we could instead look into is changing the annotation parameters of the POJOs via reflection so each test uses a unique table.

Thanks for the suggestion Chesnay ! I'll do that.

zentol commented:
> Can same tests work in parallel ?

No, a given test class (e.g., CassandraConnectorITCase) is only run in a single thread. Multiple IT cases may run at the same time, but they usually don't share resources.

echauchot commented:
> > Can same tests work in parallel ?
> 
> No, a given test class (e.g., CassandraConnectorITCase) is only run in a single thread. Multiple IT cases may run at the same time, but they usually don't share resources.

Ok thanks for the info

echauchot commented:
@zentol I've finished the annotation manipulation code through the reflection API and added a test to check that the manipulation works and rebased on master PTAL.

zentol commented on flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CustomCassandraAnnotatedPojo.java:
```diff
@@ -21,11 +21,9 @@
 
 /** Example of Cassandra Annotated POJO class for use with {@link CassandraPojoInputFormat}. */
-@Table(name = CustomCassandraAnnotatedPojo.TABLE_NAME, keyspace = "flink")
+@Table(name = "batches", keyspace = "flink")
```
we should be able to just remove this (and other) annotations because we override them anyway.

echauchot replied:
I agree thanks: as we will now add the annotation rather than modifying the existing one. It will be less confusing

zentol commented on flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:
```diff
@@ -206,6 +209,79 @@ protected Cluster buildCluster(Cluster.Builder builder) {
+        Table clazTableAnnotation = claz.getAnnotation(Table.class);
+        final String newName = clazTableAnnotation.name() + testName;
+        // BEWARE it is JDK 8 specific
+        final Method annotationDataMethod = Class.class.getDeclaredMethod("annotationData", null);
```
Have you looked into using libraries to achieve this? Like byte-buddy, asm etc.

An UNTESTED prototype:
```
    @Test
    public void testCassandraPojoAtLeastOnceSink() throws Exception {
        final String someName = "test";
        session.execute(CREATE_TABLE_QUERY.replace(TABLE_NAME_VARIABLE, someName));

        final Class<? extends Pojo> load =
                new ByteBuddy()
                        .subclass(Pojo.class)
                        .annotateType(createTableAnnotation(KEYSPACE, someName))
                        .make()
                        .load(getClass().getClassLoader())
                        .getLoaded();

        writeInstances(load);

        ResultSet rs = session.execute(SELECT_DATA_QUERY.replace(TABLE_NAME_VARIABLE, someName));
        Assert.assertEquals(20, rs.all().size());
    }

    private Table createTableAnnotation(@Nullable String keySpace, String tableName) {
        return new Table() {
            @Override
            public Class<? extends Annotation> annotationType() {
                return Table.class;
            }

            @Override
            public String keyspace() {
                return keySpace;
            }

            @Override
            public String name() {
                return tableName;
            }

            @Override
            public boolean caseSensitiveKeyspace() {
                return false;
            }

            @Override
            public boolean caseSensitiveTable() {
                return false;
            }

            @Override
            public String writeConsistency() {
                return "";
            }

            @Override
            public String readConsistency() {
                return "";
            }
        };
    }

    private <T> void writeInstances(Class<T> clazz) throws Exception {
        final Constructor<T> constructor =
                clazz.getDeclaredConstructor(String.class, Integer.class, Integer.class);
        CassandraPojoSink<T> sink = new CassandraPojoSink<>(clazz, builderForWriting);
        try {
            sink.open(new Configuration());
            for (int x = 0; x < 20; x++) {
                sink.send(constructor.newInstance(UUID.randomUUID().toString(), x, 0));
            }
        } finally {
            sink.close();
        }
    }
```

echauchot replied:
Actually my comment about java version in the code is not accurate, I should have written `JDK8+ COMPLIANT` instead of `BEWARE it is JDK 8 specific`. Indeed I tested it in JDK 11 and it works as well. I will just remove the comment as the whole code is JDK8+. But anyway, using ByteBuddy seems less error prone, I'll use it instead, thanks for the snippet.

echauchot replied:
@zentol I did the replacement of the native reflection code to ByteBuddy. But it seems that ByteBuddy needs to subclass the Pojo class to be able to manipulate its annotations. The problem with that is that the Cassandra mapper downstream no longer sees the fields defined in the super-class (Pojo) if we use ByteBuddy annotated sub-class. WDYT ?

zentol replied:
I'll tinker a bit!

echauchot replied:
Thanks Chesnay for the patch !
Sorry, you based your work on my partially updated to ByteBuddy code. In the meantime I pushed a fully migrated code. 
The big difference I see in your code is using ByteBuddy `redefine()` instead of `subClass()` that is precisely what I was looking for, thanks ! I'll update my latest code with it.

echauchot replied:
@zentol all done, PTAL

zentol commented on flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/Pojo.java:
```diff
@@ -18,26 +18,31 @@
 
     @Column(name = "counter")
-    private int counter;
+    public Integer counter;
```
Why don't you use Integer.TYPE for the constructor like I did in the prototype?

zentol replied:
I also don't think they need to be public.

echauchot replied:
I did not see you were using `Integer.TYPE` in the prototype. Fixed.

zentol commented on flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:
```diff
@@ -206,6 +205,66 @@ protected Cluster buildCluster(Cluster.Builder builder) {
+        };
+    }
+
+    @Test
```
let's re-order this a bit so we don't have a test and noisy tools in between the setup methods. Maybe move all the reflection stuff into a separate utility class.

echauchot replied:
+1

zentol commented on flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:
```diff
@@ -520,43 +575,35 @@ public void testCassandraRowAtLeastOnceSink() throws Exception {
 
-        CassandraPojoSink<PojoNoAnnotatedKeyspace> sink =
-                new CassandraPojoSink<>(PojoNoAnnotatedKeyspace.class, builderForWriting, KEYSPACE);
+    private <T> void writePojos(Class<T> annotatedPojoClass, String keyspace) throws Exception {
```
keyspace should be annotated with `@Nullable`

echauchot commented:
@zentol I addressed all your comments and squashed commits PTAL.

echauchot commented:
Thanks for review / merge Chesnay !
